docs/source/sparse.rst

```diff
@@ -7,51 +7,152 @@
-particular structure of non-zero elements in sparse arrays as well as
-for specific operations on the arrays. PyTorch supports the following
-sparse storage formats: :ref:`COO<sparse-coo-docs>`,
-:ref:`CSR<sparse-csr-docs>`, :ref:`CSC<sparse-csc-docs>`,
-:ref:`BSR<sparse-bsr-docs>`, and :ref:`BSC<sparse-bsc-docs>`.
-
-__ https://en.wikipedia.org/wiki/Sparse_matrix
+.. warning::
 
-.. note::
+  The PyTorch API of sparse tensors is in beta and may change in the near future.
+  We highly welcome feature requests, bug reports and general suggestions as Github issues.
+
+Why and when to use sparsity
+++++++++++++++++++++++++++++
+
+By default :class:`torch.Tensor` stores elements contiguously in
+physical memory leading. This leads to efficient implementations
+of various array processing algorithms that require fast access to
+elements.
+
+Now, some users might decide to represent data such as graph adjacency
+matrices, pruned weights or points clouds by Tensors whose *elements are
+mostly zero valued*. We can provide performance optimizations for these
+use cases by providing sparse storage formats.
+
+Various sparse storage formats (`such as COO, CSR/CSC, LIL, etc.`__)
+have been developed over the years. In a nutshell they all compress data by
+tracking the positions of non-zero valued elements and perhaps some,
+but *most importantly* not all, zero valued elements. We call the uncompressed
+values *specified*. They *may* be zero. In contrast *unspecified* elements
+*must* be zero.
+
+By compressing repeat zeros sparse storage formats aim to save memory
+and computational resources on various CPUs and GPUs. Especially for high
+degrees of sparsity or highly structured sparsity this can have significant
+performance implications. As such sparse storage formats are a performance
+optimization.
+
+Just like any performance optimization sparse storage formats are not
+always advantageous for every user and user environment. Details such as
+the distribution of zero valued elements, the level of sparsity and
+the user's device matter.
+
+We aim to make it easy to try these different compression techniques,
+but we do not yet provide a programmatic means of determining the best one
+for your particular application. Of course you are always welcome to
+ask for help in our Discussion forum https://dev-discuss.pytorch.org/.
+
+Functionality overview
+++++++++++++++++++++++
+
+We want it to be straightforward to construct a sparse Tensor from a
+given dense Tensor by providing conversion routines for each layout.
+
+    >>> a = torch.tensor([[0, 1.], [0, 0]])
+    >>> b = torch.tensor([[0, 2.], [3, 0]])
+    >>> a + b
+    tensor([[0., 3.],
+            [3., 0.]])
+    >>> a + b.to_sparse()
+    tensor([[0., 3.],
+            [3., 0.]])
+    >>> b.to_sparse()
+    tensor(indices=tensor([[0, 1],
+                           [1, 0]]),
+           values=tensor([2., 3.]),
+           size=(2, 2), nnz=2, layout=torch.sparse_coo)
```

Comment:
it might be clearer to just show a single `to_sparse()` conversion here. (unless I'm missing the point of the addition and the dense/sparse addition in the middle)